shell/src/main/java/org/apache/accumulo/shell/Shell.java

```diff
@@ -1181,11 +1182,14 @@ public ConsoleReader getReader() {
 
   public void updateUser(String principal, AuthenticationToken token)
       throws AccumuloException, AccumuloSecurityException {
+
     if (accumuloClient != null) {
+      if (!accumuloClient.securityOperations().authenticateUser(principal, token)) {
+        throw new AccumuloSecurityException(principal, SecurityErrorCode.BAD_CREDENTIALS);
+      }
       accumuloClient.close();
     }
     accumuloClient = Accumulo.newClient().from(clientProperties).as(principal, token).build();
-    accumuloClient.securityOperations().authenticateUser(principal, token);
     context = (ClientContext) accumuloClient;
```

Comment:
I was thinking about the case where accumuloClient is null. No authentication of the use is performed in that case. That might be okay, because that only happens on startup and there might already be an earlier failure that would prevent startup with incorrect credentials. However, I'm not sure about that without looking into it more.

An alternative implementation, that doesn't rely on assumptions about how/when this method is used is to use a local variable for the client, while it is being authenticated, and only swapping it out, if it passes authentication, as in:

```java
var newClient = Accumulo.newClient().from(clientProperties).as(principal, token).build();
newClient.securityOperations().authenticateUser(principal, token);
var oldClient = accumuloClient;
accumuloClient = newClient; // swap out old client if the new client has authenticated
context = (ClientContext) accumuloClient; // update the context with the new client
if (oldClient != null) {
  // clean up the old client
  oldClient.close();
}
```

Reply:
You bring up good points. I did think about those cases when I made this change but from what I saw, this function was only called for change user and the update password command. I could be wrong on that. I will look to implementing your suggestion as I agree that being on the safe side of things is best.

Reply:
I am running into a warning about an unreferenced accumulo client without calling close. Looking into it now.

Reply:
> I am running into a warning about an unreferenced accumulo client without calling close. Looking into it now.

Were you able to solve that with your more recent changes?

Reply:
Yes, the newClient.close() resolves this.